_posts/2020-07-21-cube-functionality-to-kartothek.markdown

```diff
@@ -0,0 +1,414 @@
+>>> print_filetree(store)
+geodata++seed.by-dataset-metadata.json
+geodata++seed/indices/city/<ts>.by-dataset-index.parquet
+geodata++seed/indices/day/<ts>.by-dataset-index.parquet
+geodata++seed/table/_common_metadata
+geodata++seed/table/country=DE/<uuid>.parquet
+geodata++seed/table/country=UK/<uuid>.parquet
+```
+
+Thus, A cube can be visualised as a metadataset of multiple datasets as shown in below image.
+ 
+ 
+![Cube Image](/assets/images/2020-07-21-kartothek-cube.png)
+
+
+Similarly, we can **Extend** this cube by **Adding** new columns to the dataframes.
+
+## Extend Operation
+
+Now let’s say we also would like to have longitude and latitude data in our cube.
+
+```python
+>>> from kartothek.io.eager_cube import extend_cube
+>>> df_location = pd.read_csv(
+...     filepath_or_buffer=StringIO("""
+...    city country  latitude  longitude
+... Hamburg      DE 53.551086   9.993682
+... Dresden      DE 51.050407  13.737262
+...  London      UK 51.509865  -0.118092
+...   Tokyo      JP 35.652832 139.839478
+...     """.strip()),
+...     delim_whitespace=True,
+... )
+```
+
+```
+>>> datasets_extend = extend_cube(
+...   data={"latlong": df_location},
+...   store=store,
+...   cube=cube,
+... )
+```
+
+This results in an extra dataset:
+
+```python
+>>> print(", ".join(sorted(datasets_extend.keys())))
+latlong
+>>> ds_latlong = datasets_extend["latlong"].load_all_indices(store)
+>>> print(ds_latlong.uuid)
+geodata++latlong
+>>> print(", ".join(sorted(ds_latlong.indices)))
+country
+```
+Note that for the second dataset, no indices for **city** and **day** exists. 
+These are only created for the seed dataset, since that datasets forms the groundtruth about which city-day entries are part of the cube.
+(Dataset that provides the groundtruth about which Cells are in a Cube is called the **seed dataset**).
+
+If you look at the file tree, you can see that the second dataset is completely separated. This is useful to copy/backup parts of the cube:
+
+```python
+>>> print_filetree(store)
+geodata++latlong.by-dataset-metadata.json
+geodata++latlong/table/_common_metadata
+geodata++latlong/table/country=DE/<uuid>.parquet
+geodata++latlong/table/country=JP/<uuid>.parquet
+geodata++latlong/table/country=UK/<uuid>.parquet
+geodata++seed.by-dataset-metadata.json
+geodata++seed/indices/city/<ts>.by-dataset-index.parquet
+geodata++seed/indices/day/<ts>.by-dataset-index.parquet
+geodata++seed/table/_common_metadata
+geodata++seed/table/country=DE/<uuid>.parquet
+geodata++seed/table/country=UK/<uuid>.parquet
+```
+## Query
+
+The whole beauty of Cube does not come from storing multiple datasets,
+but especially from retrieving the data  (**Querying**)  in a very comfortable way. 
+Kartothek views the whole cube as a large, virtual DataFrame.
+The seed dataset presents the groundtruth regarding rows, all other datasets are joined via a left join. 
+Cube naturally supports **partition-by** semantic, which is more helpful for distributed backends.
```

Comment:
Do we want to put an emphasise on this feature? We have this feature in plain kartothek as well with a different name. imho, this is not absolutely required to motivate the usage of this

Reply:
At least for the transformation section (see the other thread about this) I think this would be worth to mention.

Reply:
removed above lines.